scripts/gen_lsp.lua

```diff
@@ -154,95 +193,177 @@ function M.gen(opt)
     'decimal',
   }
 
-  local function parse_type(type)
+  ---@param documentation string
+  local _process_documentation = function(documentation)
+    documentation = documentation:gsub('\n', '\n---')
+    -- Remove <200b> (zero-width space) unicode characters: e.g., `**/<200b>*`
+    documentation = documentation:gsub('\226\128\139', '')
+    -- Escape annotations that are not recognized by lua-ls
+    documentation = documentation:gsub('%^---@sample', '---\\@sample')
+    return '---' .. documentation
+  end
+
+  --- @class vim._gen_lsp.Type
+  --- @field kind string a common field for all Types.
+  --- @field name? string for ReferenceType, BaseType
+  --- @field element? any for ArrayType
+  --- @field items? vim._gen_lsp.Type[] for OrType, AndType
+  --- @field key? vim._gen_lsp.Type for MapType
+  --- @field value? string|vim._gen_lsp.Type for StringLiteralType, MapType, StructureLiteralType
+
+  ---@param type vim._gen_lsp.Type
+  ---@param prefix? string Optional prefix associated with the this type, made of (nested) field name.
+  ---             Used to generate class name for structure literal types.
+  ---@return string
+  local function parse_type(type, prefix)
+    -- ReferenceType | BaseType
     if type.kind == 'reference' or type.kind == 'base' then
       if vim.tbl_contains(simple_types, type.name) then
         return type.name
       end
       return 'lsp.' .. type.name
+
+    -- ArrayType
     elseif type.kind == 'array' then
-      return parse_type(type.element) .. '[]'
+      return parse_type(type.element, prefix) .. '[]'
+
+    -- OrType
     elseif type.kind == 'or' then
       local val = ''
       for _, item in ipairs(type.items) do
-        val = val .. parse_type(item) .. '|'
+        val = val .. parse_type(item, prefix) .. '|' --[[ @as string ]]
       end
       val = val:sub(0, -2)
       return val
+
+    -- StringLiteralType
     elseif type.kind == 'stringLiteral' then
       return '"' .. type.value .. '"'
+
+    -- MapType
     elseif type.kind == 'map' then
-      return 'table<' .. parse_type(type.key) .. ', ' .. parse_type(type.value) .. '>'
+      local key = assert(type.key)
+      local value = type.value --[[ @as vim._gen_lsp.Type ]]
+      return 'table<' .. parse_type(key, prefix) .. ', ' .. parse_type(value, prefix) .. '>'
+
+    -- StructureLiteralType
     elseif type.kind == 'literal' then
       -- can I use ---@param disabled? {reason: string}
       -- use | to continue the inline class to be able to add docs
       -- https://github.com/LuaLS/lua-language-server/issues/2128
       anonymous_num = anonymous_num + 1
-      local anonym = { '---@class anonym' .. anonymous_num }
-      for _, field in ipairs(type.value.properties) do
+      local anonymous_classname = 'lsp._anonym' .. anonymous_num
```

Comment:
Would it cause unintended conflicts without the `_anonym`+number part?

I see there's:

```
@class lsp._anonym50.TextDocumentFilter
@class lsp._anonym49.TextDocumentFilter
```

I suspect if we've functions that deal with these properties it might be more useful if they were actually the same type - given that they are the same thing.

Reply:
That's a great point I also noticed; I was planning to have a further improvement, but possibly as a separate PR (it's going to be a complicated one).

Eventually, we should remove the "anonym" thing; in this particular case the class can be named as something like `lsp.TextDocumentFilter`.

In most of such cases where several structured literal types (with the same filed name) are involved, they only differ in whether a field is optional or not. For example:

```diff
 ---@class lsp._anonym49.TextDocumentFilter
+---@field language string
 ---@field scheme? string
 ---@field pattern? string

 ---@class lsp._anonym50.TextDocumentFilter
 ---@field language? string
+---@field scheme string
 ---@field pattern? string

 ---@class lsp._anonym51.TextDocumentFilter
 ---@field language? string
 ---@field scheme? string
+---@field pattern string
```

The semantic here is either `language`, `scheme`, or `pattern` is given (but at least only one). This could be consolidated into one class:

```
---@class lsp.TextDocumentFilter
---@field language? string
---@field scheme? string
---@field pattern? string
```

since the "exclusive or" semantic here is not very important for us. In principle we can do some sort of type unification and merge types if applicable.